README.md
Proxy Checker
=============

Check if you proxy is working or not very easily

## Installation

If you don't want to compile your own version, you can use the following repository to install it 

### Debian

```bash
echo "deb http://packages.matoski.com/ debian main" | sudo tee /etc/apt/sources.list.d/packages-matoski-com.list
curl -s http://packages.matoski.com/keyring.gpg | sudo apt-key add -
sudo apt-get update
sudo apt-get install proxy-checker
```

## Getting Started with Proxy Checker

### Requirements

* [Golang](https://golang.org/dl/) >= 1.8
* [Glide](https://github.com/Masterminds/glide) >= 0.12.3

### Autocomplete

You know what to do with this

* [Bash](contrib/proxy-checker.bash)
* [Zsh](contrib/proxy-checker.zsh)

### Dependencies 

This project uses glide to manage dependencies so download them before trying to build/install by running 

```bash
glide install
```

### Build

To build the binary for Proxy Checker run the command below. This will generate a binary
in the bin directory with the name proxy-checker.

```bash
make build
```

### Install

To install the binary for Proxy Checker run the command below. This will generate a binary
in $GOPATH/bin directory with the name proxy-checker and add the bash autocomplete files.

```bash
make install
```

### CSV file

The csv file structure is like this, you can add as many as you want

```csv
<schema>://<host>[:<port>],<username>,<password>
```

* **\<schema>** is http, or https
* **\<host>** is the actually host of the proxy, or it can even be an IP address
* **\<port>** is the port of the proxy, if not supplied it defaults to 80
* **\<username>** is the username for the proxy
* **\<password>** is the password for the proxy

## Run

### Help
```bash
$ proxy-checker --help
usage: proxy-checker [<flags>] <command> [<args> ...]

Checks if an http proxy with basic auth works by querying https://api.ipify.org/

Flags:
  -h, --help         Show context-sensitive help (also try --help-long and --help-man).
      --queue=25     How many request to process at one time
      --failed-only  Show only failed proxies
      --host="https://api.ipify.org"  
                     Host to query for checking the proxy

Commands:
  help [<command>...]
    Show help.

  version
    Show version and terminate

  check <host-port> [<username>] [<password>]
    Check the single proxy

  csv-file <file>
    Check all the proxies in the file specified
```

### Single proxy

```bash
proxy-checker check http://example.com:60000 username password
http://username:password@example.com:60000 is valid
```

### Checking multiple proxies

Where **proxies.csv** contains

```csv
http://proxy01.com:60000,username1,password1
http://proxy02.com:9000,username2,password2
http://proxy03.com:14000,username3,password3
http://proxy04.com:60000,username4,password4
http://proxy05.com,username5,password5
```

And you can easily check all of them with this command

```bash
proxy-checker csv-file proxies.csv
http://username1:password1@proxy01.com:60000 is valid
http://username2:password2@proxy02.com:9000 is invalid
http://username3:password3@proxy03.com:14000 is invalid
http://username4:password4@proxy04.com:60000 is valid
http://username5:password5@proxy05.com is valid
```

## License

Apache License
Version 2.0, January 2004

See [License](LICENSE) file


debian/changelog
proxy-checker (0.1.4) unstable; urgency=medium

  * Compiled with golang 1.8, the problem with the http.Client is fixed, removed --use-lock

 -- Ilija Matoski <ilijamt@gmail.com>  Tue, 21 Feb 2017 19:55:43 +0100

proxy-checker (0.1.3) unstable; urgency=medium

  * Added a property to specify the URL to fetch, which should return the status code and content length

 -- Ilija Matoski <ilijamt@gmail.com>  Tue, 21 Feb 2017 08:28:23 +0100

proxy-checker (0.1.1) unstable; urgency=medium

  * Could not print the version of the application.

 -- Ilija Matoski <ilijamt@gmail.com>  Thu, 13 Oct 2016 08:57:20 +0200

proxy-checker (0.1.0) unstable; urgency=low

  * Initial release.
  * Check a single proxy.
  * Check multiple proxies from a csv file.

 -- Ilija Matoski <ilijamt@gmail.com>  Wed, 12 Oct 2016 22:17:43 +0200


job/job_test.go
package job

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func Shim(a, b interface{}) []interface{} {
	return []interface{}{a, b}
}

func Test_DetailWithNoAuthenticationDetails(t *testing.T) {
	var item = Detail{
		Host: "http://example.com:9000",
	}
	var url, err = item.ToURL()
	assert.Nil(t, err)
	assert.NotNil(t, url)
	assert.EqualValues(t, "example.com:9000", url.Host)
	assert.EqualValues(t, "http", url.Scheme)
}

func Test_DetailWithIncompleteAuthenticationDetails(t *testing.T) {
	var item = Detail{
		Host:     "http://example.com:9000",
		Username: "user",
	}
	var url, err = item.ToURL()
	assert.Nil(t, err)
	assert.NotNil(t, url)
	assert.EqualValues(t, "example.com:9000", url.Host)
	assert.EqualValues(t, "http", url.Scheme)
	assert.Nil(t, url.User)
}

func Test_DetailWithAuthenticationDetails(t *testing.T) {
	var item = Detail{
		Host:     "http://example.com:9000",
		Username: "user",
		Password: "pass",
	}
	var url, err = item.ToURL()
	assert.Nil(t, err)
	assert.NotNil(t, url)
	assert.EqualValues(t, "example.com:9000", url.Host)
	assert.EqualValues(t, "http", url.Scheme)
	assert.NotNil(t, url.User)
	assert.EqualValues(t, "http://user:pass@example.com:9000", url.String())
}


job/queue.go
package job

import (
	"fmt"
	"net/http"
	"net/url"
	"sync"
	"time"
	"context"
	"io/ioutil"
)

type Queue struct {
	jobs chan Detail
	quit chan bool

	queryURL *url.URL

	failedOnly bool

	wg   sync.WaitGroup
}

func (q *Queue) Wait() {
	q.wg.Wait()
}

func NewQueue(queueSize int, host string, failedOnly bool) *Queue {
	queue := &Queue{}
	queue.jobs = make(chan Detail, queueSize)
	queue.quit = make(chan bool, 1)
	queue.queryURL, _ = url.Parse(host)
	queue.failedOnly = failedOnly
	return queue
}

func (q *Queue) Run() {
	var quit bool = false

	for {
		select {

		// we got a new job to process
		case j := <-q.jobs:
			go q.IsValidProxy(j)

		// we got the signal to quit
		case <-q.quit:
			quit = true
		}

		// quit if we have finished
		if quit {
			break
		}
	}
}

func (q *Queue) Stop() {
	q.quit <- true
}

func (q *Queue) IsValidProxy(job Detail) bool {
	defer q.wg.Done()

	var err error
	var validStatus bool = false
	var response *http.Response
	var proxyURL, _ = job.ToURL()

	defer func() {
		var rsp = "invalid"
		if validStatus {
			rsp = "valid"
		}

		var status string
		var contentLength int64

		if response == nil {
			status = "UNKNOWN"
			contentLength = 0
		} else {
			status = response.Status
			contentLength = response.ContentLength
			if contentLength <= 0 {
				body, _ := ioutil.ReadAll(response.Body)
				l := len(body)
				contentLength = int64(l)
			}
		}

		if q.failedOnly {
			if !validStatus {
				fmt.Printf("%s is %s (Status Code: %s, Content Length: %d)\n", proxyURL.String(), rsp, status, contentLength)
			}
		} else {
			fmt.Printf("%s is %s (Status Code: %s, Content Length: %d)\n", proxyURL.String(), rsp, status, contentLength)
			if validStatus {
				ioutil.ReadAll(response.Body)
			}
		}

	}()

	ctx, _ := context.WithTimeout(context.Background(), 5*time.Second)
	client := &http.Client{
		Transport: &http.Transport{
			Proxy: http.ProxyURL(proxyURL),
		},
	}
	req, _ := http.NewRequest("GET", q.queryURL.String(), nil)
	req = req.WithContext(ctx)
	response, err = client.Do(req)

	validStatus = err == nil
	if response != nil {
		validStatus = err == nil && response.Status == "200 OK"
	}

	return validStatus
}

func (q *Queue) WgIncr() {
	q.wg.Add(1)
}

func (q *Queue) AddJob(job Detail) {
	q.jobs <- job
}


main.go
package main

import (
	"bufio"
	"encoding/csv"
	"fmt"
	"os"
	"runtime"

	"github.com/alecthomas/kingpin"

	"github.com/ilijamt/proxy-checker/job"
)

var (
	command string

	app        = kingpin.New(Name, Description)
	version    = app.Command("version", "Show version and terminate").Action(ShowVersion)
	queueSize  = app.Flag("queue", "How many request to process at one time").Default("25").Int()
	failedOnly = app.Flag("failed-only", "Show only failed proxies").Bool()
	host       = app.Flag("host", "Host to query for checking the proxy").Default("https://api.ipify.org").String()

	check         = app.Command("check", "Check the single proxy")
	checkHostPort = check.Arg("host-port", "The hostname of the proxy with the proxy, add the schema to the URL like https://proxy.com:9000").Required().String()
	checkUsername = check.Arg("username", "The username to use for the proxy, can be empty if no authentication is required").String()
	checkPassword = check.Arg("password", "The password to use for the proxy, can be empty if no authentication is required").String()

	file     = app.Command("csv-file", "Check all the proxies in the file specified")
	fileName = file.Arg("file", "The file name to load").Required().File()

	queue *job.Queue
)

func ShowVersion(c *kingpin.ParseContext) error {
	fmt.Printf("%s version %s build %s (%s), built on %s, by %s\n", Name, BuildVersion, BuildHash, runtime.GOARCH, BuildDate, Maintainer)
	os.Exit(0)
	return nil
}

func init() {

	app.HelpFlag.Short('h')
	command = kingpin.MustParse(app.Parse(os.Args[1:]))

	queue = job.NewQueue(*queueSize, *host, *failedOnly)

}

func main() {

	defer queue.Wait()
	go queue.Run()

	switch command {
	case "check":

		queue.WgIncr()
		go queue.AddJob(job.Detail{
			Host:     *checkHostPort,
			Username: *checkUsername,
			Password: *checkPassword,
		})

	case "csv-file":
		f, e := os.Open((*fileName).Name())

		if e != nil {
			_ = fmt.Errorf("%v", e)
			return
		}

		r := csv.NewReader(bufio.NewReader(f))
		result, _ := r.ReadAll()

		var host string
		var username string
		var password string

		for i := range result {

			host = result[i][0]
			username = result[i][1]
			password = result[i][2]

			queue.WgIncr()
			go queue.AddJob(job.Detail{
				Host:     host,
				Username: username,
				Password: password,
			})

		}

	}

}
